tests/subsys/audio_module/src/functional_test.c

```diff
@@ -0,0 +1,1801 @@
+			      ret);
+		zassert_equal(data_fifo_empty_fake.call_count, 0,
+			      "Failed close, data FIFO empty called %d times",
+			      data_fifo_empty_fake.call_count);
+		zassert_mem_equal(&mod_description, &test_mod_description,
+				  sizeof(struct audio_module_description),
+				  "Failed close, modified the modules description");
+		zassert_mem_equal(&mod_functions_null, &test_mod_functions,
+				  sizeof(struct audio_module_functions),
+				  "Failed close, modified the modules functions");
+		zassert_mem_equal(&mod_parameters, &test_mod_parameters,
+				  sizeof(struct audio_module_parameters),
+				  "Failed close, modified the modules parameter settings");
+		zassert_mem_equal(&mod_context, &mod_context, sizeof(struct mod_context),
+				  "Failed close, modified the modules context");
+
+		memset(&handle, 0, sizeof(struct audio_module_handle));
+		memcpy(&handle.name, test_inst_name, sizeof(test_inst_name));
+		handle.description = &mod_description;
+		handle.use_tx_queue = true;
+		handle.dest_count = TEST_MODULES_NUM;
+		handle.thread.stack = mod_stack;
+		handle.thread.stack_size = TEST_MOD_THREAD_STACK_SIZE;
+		handle.thread.priority = TEST_MOD_THREAD_PRIORITY;
+		handle.thread.data_slab = &mod_data_slab;
+		handle.thread.data_size = TEST_MOD_DATA_SIZE;
+		handle.context = (struct audio_module_context *)&mod_context;
+
+		/* Fake internal empty data FIFO success */
+		data_fifo_empty_fake.custom_fake = fake_data_fifo_empty__succeeds;
+
+		handle.thread.msg_rx = NULL;
+		handle.thread.msg_tx = NULL;
+
+		start_thread(&handle);
+
+		handle.state = AUDIO_MODULE_STATE_STOPPED;
+
+		ret = audio_module_close(&handle);
+
+		zassert_equal(ret, 0, "Close function did not return successfully (0): ret %d",
+			      ret);
+		zassert_equal(data_fifo_empty_fake.call_count, 0,
+			      "Failed close, data FIFO empty called %d times",
+			      data_fifo_empty_fake.call_count);
+		zassert_mem_equal(&mod_description, &test_mod_description,
+				  sizeof(struct audio_module_description),
+				  "Failed close, modified the modules description");
+		zassert_mem_equal(&mod_functions_null, &test_mod_functions,
+				  sizeof(struct audio_module_functions),
+				  "Failed close, modified the modules functions");
+		zassert_mem_equal(&mod_parameters, &test_mod_parameters,
+				  sizeof(struct audio_module_parameters),
+				  "Failed close, modified the modules parameter settings");
+		zassert_mem_equal(&mod_context, &mod_context, sizeof(struct mod_context),
+				  "Failed close, modified the modules context");
+	}
+}
+
+ZTEST(suite_audio_module_functional, test_close_fnct)
+{
+	int ret;
+	char *test_inst_name = "TEST instance 1";
+	struct audio_module_description test_mod_description;
+	struct mod_context test_mod_context;
+	struct audio_module_functions test_mod_functions = mod_functions_null;
+
+	mod_description.functions = &mod_functions_null;
+
+	for (int i = AUDIO_MODULE_TYPE_INPUT; i <= AUDIO_MODULE_TYPE_IN_OUT; i++) {
```

Comment:
Discuss: Many of these tests test across different module types as well as different states.

Reply:
Yes, this is tricky without repeating the test for each type and adding code. Also, I would hope the test would fail if a new type was added. This would benefit from a macro passing in each valid type.